pkg/cloud/services/s3/s3.go

```diff
@@ -0,0 +1,249 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package s3
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"path"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/aws/aws-sdk-go/service/sts/stsiface"
+	"github.com/pkg/errors"
+
+	iam "sigs.k8s.io/cluster-api-provider-aws/iam/api/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/scope"
+)
+
+// Service holds a collection of interfaces.
+// The interfaces are broken down like this to group functions together.
+// One alternative is to have a large list of functions from the ec2 client.
+type Service struct {
+	scope     scope.S3Scope
+	S3Client  s3iface.S3API
+	STSClient stsiface.STSAPI
+}
+
+// NewService returns a new service given the api clients.
+func NewService(s3Scope scope.S3Scope) *Service {
+	s3Client := scope.NewS3Client(s3Scope, s3Scope, s3Scope, s3Scope.InfraCluster())
+	STSClient := scope.NewSTSClient(s3Scope, s3Scope, s3Scope, s3Scope.InfraCluster())
+
+	return &Service{
+		scope:     s3Scope,
+		S3Client:  s3Client,
+		STSClient: STSClient,
+	}
+}
+
+func (s *Service) ReconcileBucket() error {
+	if !s.bucketManagementEnabled() {
+		return nil
+	}
+
+	bucketName := s.bucketName()
+
+	if err := s.createBucketIfNotExist(bucketName); err != nil {
+		return errors.Wrap(err, "ensuring bucket exists")
+	}
+
+	if err := s.ensureBucketPolicy(bucketName); err != nil {
+		return errors.Wrap(err, "ensuring bucket policy")
+	}
+
+	return nil
+}
+
+func (s *Service) DeleteBucket() error {
+	if !s.bucketManagementEnabled() {
+		return nil
+	}
+
+	bucketName := s.bucketName()
+
+	s.scope.Info("Deleting S3 Bucket", "name", bucketName)
+
+	if _, err := s.S3Client.DeleteBucket(&s3.DeleteBucketInput{
+		Bucket: aws.String(bucketName),
+	}); err != nil {
+		return errors.Wrap(err, "deleting S3 bucket")
+	}
+
+	return nil
+}
+
+func (s *Service) Create(m *scope.MachineScope, data []byte) (string, error) {
+	if !s.bucketManagementEnabled() {
+		return "", errors.New("requested object creation but bucket management is not enabled")
+	}
+
+	if m == nil {
+		return "", errors.New("machine scope can't be nil")
+	}
+
+	if len(data) == 0 {
+		return "", errors.New("got empty data")
+	}
+
+	bucket := s.bucketName()
+	key := s.bootstrapDataKey(m)
+
+	if _, err := s.S3Client.PutObject(&s3.PutObjectInput{
+		Body:                 aws.ReadSeekCloser(bytes.NewReader(data)),
+		Bucket:               aws.String(bucket),
+		Key:                  aws.String(key),
+		ServerSideEncryption: aws.String("aws:kms"),
+	}); err != nil {
+		return "", errors.Wrap(err, "putting object")
+	}
+
+	objectURL := &url.URL{
+		Scheme: "s3",
+		Host:   bucket,
+		Path:   key,
+	}
+
+	return objectURL.String(), nil
+}
+
+func (s *Service) Delete(m *scope.MachineScope) error {
```

Comment:
If same bucket is re-used by multiple clusters, we should perhaps make sure we don't remove the bucket with existing objects OR print a warning when there are remaining objects in the bucket and bucket has not been removed.

Reply:
Done.